### PR TITLE
Refactor Alice's revoke method to accept a `TreasureMap`

### PR DESCRIPTION
**Pull request by tuxxy**
### What this does:

1. Refactors `Alice.revoke` to take a `TreasureMap` affiliated with an active policy instead of a `Policy` directly.
2. The `RevocationKit` now also takes a ` TreasureMap` instead of a `Policy`.
3. Removes `add_active_policy` and `active_policies`, as well as the `revocation_kits` dict.
4. Adds two methods `__serialize` and `__deserialize` to `TreasureMap`.
    - This is not super fine work, but it's a hacky way to get going in the direction that we eventually need to be in.
5. The `grant` character control method additionally returns a `PolicyCredential` for Alice (or Bob) to use as needed.

### CODE REVIEW

File: nucypher/characters/lawful.py [hunk 1]
```diff
@@ -292,28 +279,30 @@ def get_policy_encrypting_key_from_label(self, label: bytes) -> UmbralPublicKey:
         policy_pubkey = alice_delegating_power.get_pubkey_from_label(label)
         return policy_pubkey
 
-    def revoke(self, policy) -> Dict:
+    def revoke(self, treasure_map) -> Dict:
```
Comment (michwill) [inline nucypher/characters/lawful.py]: So, do we want want to revoke by treasure_map or by revocation_kit?

Comment (michwill) [reply nucypher/characters/lawful.py]: Ah, I see, it's probably better to keep only the minimum needed here indeed

File: nucypher/policy/collections.py [hunk 2]
```diff
@@ -163,6 +167,29 @@ def public_id(self):
         _id = keccak_digest(bytes(self._verifying_key) + bytes(self._hrac)).hex()
         return _id
 
+    def __serialize(self):
+        """
+        WAIT! This is probably not what you want. In 9/10 times, you'll
```
Comment (KPrasch) [inline nucypher/policy/collections.py]: It's not clear what's happening here to me - What are the differences between this and the other serilizaers?

Comment (michwill) [reply nucypher/policy/collections.py]: The "normal" serializer also encrypts for Bob (yeah, very convoluted logic). This one serializes a plaintext TreasureMap

**Comment by michwill** [comment]
> Help me understand at a high-level - What is the motivation for making this change?

Basically, Alice may not have enough information at the time she revokes to construct a whole policy object. But you only need the treasure map to actually revoke.

I do agree though that it's logical to revoke by policy. Maybe we can have a separate method?

File: nucypher/policy/collections.py (hunk 2 above)
Comment (tuxxy) [reply nucypher/policy/collections.py]: Yes, essentially, when you call `bytes(...)` on a `TreasureMap`, it assumes you want it encrypted for Bob. Attempting to change this causes massive ripple effects.

**Comment by tuxxy** [comment]
@KPrasch From an API design standpoint, it may be more clear _by name only_.

Once you start digging into the internals of `Policy`, it's really obvious that using it, serializing it, and passing it around places is exceptionally burdensome. It's becoming far more apparent that `Policy` needs some refactoring to get it in a place where it may be useful for anything like this. As of right now, it feels extremely inappropriate to use a `Policy` at an API level, unfortunately.

When we start reducing what we _need_ from a `Policy` to use in the API, we essentially construct a `PolicyCredential`. Though, this may be evolving into as @jMyles a parent class.

**Comment by michwill** [comment]
Hmm, would it actually revoke on chain?

File: nucypher/policy/collections.py [hunk 3]
```diff
@@ -163,6 +167,29 @@ def public_id(self):
...
+        probably want to use `bytes(my_treasure_map)`. So... do that.
+        """
+        return self._public_signature + self._hrac + self._m.to_bytes(1, 'big') + self.nodes_as_bytes()
+
+    @classmethod
+    def __deserialize(cls, bytes_representation):
+        """
+        WAIT! This is probably not what you want. In 9/10 times, you'll probably
```
Comment (KPrasch) [inline nucypher/policy/collections.py]: Hmm - can you be more specific bout the condition in which you *do* want to use this method in the doctring?

File: tests/characters/control/test_web_control.py [hunk 4]
```diff
@@ -99,12 +99,15 @@ def test_alice_character_control_revoke(alice_web_controller_test_client, federa
     response = alice_web_controller_test_client.put('/grant', data=json.dumps(grant_request_data))
     assert response.status_code == 200
 
+    policy_credential = PolicyCredential.from_json(
+                                response.json['result']['policy_credential'])
```
Comment (KPrasch) [inline tests/characters/control/test_web_control.py]: This style of formatting (odd newline placement, IMO) is throughout your PR - are running black?

Comment (michwill) [reply tests/characters/control/test_web_control.py]: That's the way to please PEP8 gods with the 80-character rule!

Comment (KPrasch) [reply tests/characters/control/test_web_control.py]: in this case (and in many) you can get a shorter, more readable block by adding a variable:

```
    credential_json = response.json['result']['policy_credential']
    policy_credential = PolicyCredential.from_json()
```

**Comment by KPrasch** [comment]
Included in #1367

**Comment by tuxxy** [comment]
Closing in favor of #1367

**Pull request closed without merging**